### ISSUE TITLE
Question for TShader::addProcesses(const std::vector<std::string>&)

### ISSUE THREAD

**Issue by lch32111** [KhronosGroup/glslang#2163]
Hi. I'd like to say "Thank you for maintaining this good open source" in advance.

I want to ask how to use this method of TShader Class :
https://github.com/KhronosGroup/glslang/blob/bfe4c5957fc51148a0aab6e04bb22020667c1092/glslang/Public/ShaderLang.h#L441

I can see some comments of how to use a `TShader` class, but there is no description for addProcesses. I was trying to use setPreamble and referenced the StandAlone.cpp code for it.
https://github.com/KhronosGroup/glslang/blob/bfe4c5957fc51148a0aab6e04bb22020667c1092/StandAlone/StandAlone.cpp#L195

after I saw the code, I guess it's for the spv conversion from glsl code. But I don't have time to inspect all of internal codes. So Can anyone tell me the purpose of the method `addProcesses`?

Thanks.

**Comment by johnkslang** [comment]
You're welcome.

`addProcess()` is for generating the SPIR-V **OpModuleProcessed**-like information.  The data structure had a comment:

```
// Track a set of strings describing how the module was processed.
// Using the form:
//   process arg0 arg1 arg2 ...
//   process arg0 arg1 arg2 ...
// where everything is textual, and there can be zero or more arguments
class TProcesses {
```

In the cases that this proceeds to generate SPIR-V, it gets emitted either as **OpModuleProcessed** or as comments in versions predating that.

I will improve the comments now.

**Comment by johnkslang** [comment]
Oh, the comments for `TShader` included:

```
//  - optionally use setPreamble() to set a special shader string that will be
//    processed before all others but won't affect the validity of #version
```

The preamble is things like command-line -DFOO, where preprocessing macros get set before the first line of explicit code.

**Comment by lch32111** [comment]
Thank you @johnkslang  for your detailed answer!

So the `addProcesses` method is for the readable spir-v code in order to make a user read the spir-v code with preprocessor code. and I was able to know the purpose of the preamble with the comments and example codes. But the `addProcesses` is the one which is hard to understand, because I'm not used to the spir-v language. After searching for the `OpModuleProcessed`, I got to know the purpose of addProcesses.
 
I think Improving the comments will help novice users to use glslang easily as you said.
